autoprecompiles/src/low_degree_bus_interaction_optimizer.rs

```diff
@@ -0,0 +1,554 @@
+                    } else {
+                        None
+                    }
+                })
+            })
+            .try_fold(1u64, |acc, x| acc.checked_mul(x?))
+            .is_some_and(|count| count <= MAX_DOMAIN_SIZE)
+    }
+
+    /// Given a bus interaction and a symbolic input-output pair, tries to find a low-degree function
+    /// by testing all of the hard-coded hypotheses against set of all concrete input-output pairs.
+    fn find_low_degree_function(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+        symbolic_function: &SymbolicFunction<T, V>,
+    ) -> Option<LowDegreeFunction<T, V>> {
+        let mut hypotheses = hypotheses(symbolic_function.inputs.len());
+        let all_possible_assignments =
+            self.concrete_input_output_pairs(bus_interaction, symbolic_function);
+
+        for assignment in all_possible_assignments {
+            let Ok((inputs, output)) = assignment else {
+                // We can't enumerate all possible assignments, so the hypotheses can't be tested.
+                return None;
+            };
+            let inputs = inputs
+                .into_iter()
+                .map(|value| GroupedExpression::from_number(value))
+                .collect::<Vec<_>>();
+            hypotheses.retain(|hypothesis| {
+                let hypothesis_evaluation = hypothesis(inputs.clone());
+                hypothesis_evaluation.try_to_number().unwrap() == output
+            });
+            if hypotheses.is_empty() {
+                // No hypothesis left
+                return None;
+            }
+        }
+
+        // If we got this far, the hypothesis is correct!
+        Some(hypotheses.into_iter().exactly_one().unwrap_or_else(|_| {
+            panic!("Expected exactly one multilinear extension, but got multiple.")
+        }))
+    }
+
+    /// Returns the range constraints enforced by the given bus interaction.
+    fn range_constraints(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> RangeConstraints<T, V> {
+        let range_constraints = self
+            .bus_interaction_handler
+            .handle_bus_interaction(bus_interaction.to_range_constraints(&NoRangeConstraints));
+        bus_interaction
+            .payload
+            .iter()
+            .zip_eq(range_constraints.payload)
+            .map(|(expr, rc)| (expr.clone(), rc))
+            .filter(|(expr, _rc)| expr.try_to_number().is_none())
+            .collect()
+    }
+
+    /// Generate all concrete input-output pairs given a symbolic one.
+    ///
+    /// The inputs are generated as the cross product of all allowed values of the
+    /// individual inputs.
+    /// The outputs are generated by asking the bus interaction handler for each input assignment.
+    ///
+    /// If at any time (1) the inputs violate a constraint or (2) the outputs are not unique,
```

Comment:
It is only an error if there is some input that satisfies the range constraints such that there is not exactly one output , right?

Reply:
I'm not sure I understand the question!

This comment is referring to the two question marks below. Let me give a counter example for each:
1. Say there is one input with range constraint `0xff`, and the lookup maps all bytes *except* `123` to itself (`123` is just not in the precomputed table). Then, the lookup implements `f(x) = x`, but on a domain that is smaller than what we can represent with a simple range constraint. Then, we don't want to replace the bus interaction! (Arguably, we could still add the `x = y` constraint, which would at least lead to one column being substituted.)
2. I think this is the case you have in mind, the output is just not determined by the input.